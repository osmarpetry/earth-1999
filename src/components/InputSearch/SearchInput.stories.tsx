import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import InputSearch, { SearchInputProps } from 'components/InputSearch';

export default {
  title: 'Components/InputSearch',
  component: InputSearch,
} as Meta;

const Template: Story<SearchInputProps> = (args) => <InputSearch {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  label: 'Buscar her贸i',
  name: 'seach',
};

export const Sugestions = Template.bind({});
Sugestions.args = {
  label: 'Buscar her贸i',
  name: 'seach',
  value: 'hul',
  isSugestionsOpen: true,
  sugestions: ['hul', 'Hulk', 'She-Hulk']
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'Buscar her贸i',
  name: 'seach',
  isSecondary: true
};

export const SugestionsSecondary = Template.bind({});
SugestionsSecondary.args = {
  label: 'Buscar her贸i',
  name: 'seach',
  value: 'hul',
  isSugestionsOpen: true,
  sugestions: ['hul', 'Hulk', 'She-Hulk'],
  isSecondary: true
};
