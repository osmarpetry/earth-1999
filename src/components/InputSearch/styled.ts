import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

import colors from 'core/assets/styles/colors';
import fonts from 'core/assets/styles/fonts';
import responsive from 'core/assets/styles/responsive';
import sizes from 'core/assets/styles/sizes';

export const InputWrapper = styled.div<{ isSecondary: boolean }>`
  display: flex;
  align-items: center;
  position: relative;

  input {
    width: 100%;
    height: 100%;
    padding-left: 60px;
    font-size: ${sizes.paragraphSize};
  }

  img {
    position: absolute;
    left: 20px;
    pointer-events: none;
  }

  label {
    position: absolute;
    color: ${colors.primary};
    font-weight: ${fonts.lightBold};
    left: 60px;
    font-size: ${sizes.paragraphSize};
    transition: all 0.3s;
    pointer-events: none;
  }

  input::placeholder {
    color: transparent;
    opacity: 0.6;
    transition: 0.3s color;
    font-size: ${sizes.paragraphSize};
  }

  input:focus::placeholder {
    color: ${colors.primary};
  }

  input:focus + label {
    transform: translate(-35px, -35px);

    @media only screen and (max-width: ${responsive.mobile}) {
      transform: translate(-58px, -35px);
    }
    font-size: 2rem;
    color: ${colors.primary};
    white-space: nowrap;
  }

  input:not(:placeholder-shown) + label {
    transform: translate(-35px, -35px);

    @media only screen and (max-width: ${responsive.mobile}) {
      transform: translate(-58px, -35px);
    }
    font-size: 2rem;
    color: ${colors.primary};
    white-space: nowrap;
  }
`;

export const StyledInput = styled.input<{
  isSecondary: boolean;
  isSugestionsOpen: boolean;
}>`
  border-radius: ${({ isSugestionsOpen }) =>
    isSugestionsOpen ? '15px 15px 0 0' : '15px'};
  outline: unset;
  border: none;
  background-color: ${({ isSecondary }) => (isSecondary ? '#ffff' : '#fdecec')};
  padding: 15px;
  color: ${colors.primary};
  font-weight: ${fonts.lightBold};
`;

export  const SugestionsList = styled.ul<{ isSecondary: boolean }>`
  background: ${({ isSecondary }) => (isSecondary ? '#ffff' : '#fdecec')};
  border-radius: 0 0 15px 10px;
  margin: 0;
  padding: 0;
  list-style: none;
  z-index: 99;

  li {
    border-top: 1px solid ${colors.primary};
    padding: 10px 0;
    text-align: center;
  }
`;

export  const SugestionItemStyle = css`
  width: 100%;
  height: 100%;
  font-size: ${sizes.paragraphSize};
`;

export const SugestionButtonItem = styled(Link)`
  ${SugestionItemStyle}
`;

export  const SugestionItemStyleSpan = styled.span`
  ${SugestionItemStyle}
`;