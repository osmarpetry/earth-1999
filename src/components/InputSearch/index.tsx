import React from 'react';
import { Link } from 'react-router-dom';
import styled, { css } from 'styled-components';

import { AllHeroes } from 'core/utils/heroes';

import colors from 'core/assets/styles/colors';
import sizes from 'core/assets/styles/sizes';
import fonts from 'core/assets/styles/fonts';

import SearchIcon from 'assets/busca/Lupa/Shape@1,5x.svg';

const InputWrapper = styled.div<{ isSecondary: boolean }>`
  display: flex;
  align-items: center;
  position: relative;

  input {
    width: 100%;
    height: 100%;
    padding-left: 60px;
    font-size: ${sizes.paragraphSize};
  }

  img {
    position: absolute;
    left: 20px;
    pointer-events: none;
  }

  label {
    position: absolute;
    color: ${colors.primary};
    font-weight: ${fonts.lightBold};
    left: 60px;
    font-size: ${sizes.paragraphSize};
    transition: all 0.3s;
    pointer-events: none;
  }

  input::placeholder {
    color: transparent;
    opacity: 0.6;
    transition: 0.3s color;
    font-size: ${sizes.paragraphSize};
  }

  input:focus::placeholder {
    color: ${colors.primary};
  }

  input:focus + label {
    transform: translate(-35px, -35px);
    font-size: 2rem;
    color: ${colors.primary};
  }

  input:not(:placeholder-shown) + label {
    transform: translate(-35px, -35px);
    font-size: 2rem;
    color: ${colors.primary};
    white-space: nowrap;
  }
`;

const StyledInput = styled.input<{
  isSecondary: boolean;
  isSugestionsOpen: boolean;
}>`
  border-radius: ${({ isSugestionsOpen }) =>
    isSugestionsOpen ? '15px 15px 0 0' : '15px'};
  outline: unset;
  border: none;
  background-color: ${({ isSecondary }) => (isSecondary ? '#ffff' : '#fdecec')};
  padding: 15px;
  color: ${colors.primary};
  font-weight: ${fonts.lightBold};
`;

export interface SearchInputProps {
  isSecondary?: boolean;
  isSugestionsOpen?: boolean;
  label: string;
  placeholder: string
  name?: string;
  sugestions?: AllHeroes[];
  value: string;
  onChange?: (value: string) => void;
  onSugestionClick?: (sugestions: string) => void;
}

const SugestionsList = styled.ul<{ isSecondary: boolean }>`
  background: ${({ isSecondary }) => (isSecondary ? '#ffff' : '#fdecec')};
  border-radius: 0 0 15px 10px;
  margin: 0;
  padding: 0;
  list-style: none;
  z-index: 99;

  li {
    border-top: 1px solid ${colors.primary};
    padding: 10px 0;
    text-align: center;
  }
`;

const SugestionItemStyle = css`
  width: 100%;
  height: 100%;
  font-size: ${sizes.paragraphSize};
`;

const SugestionButtonItem = styled(Link)`
  ${SugestionItemStyle}
`;

const SugestionItemStyleSpan = styled.span`
  ${SugestionItemStyle}
`;

export default function SearchInput({
  isSecondary = false,
  isSugestionsOpen = false,
  label,
  placeholder,
  name = 'search',
  sugestions,
  onChange,
  onSugestionClick,
}: SearchInputProps) {
  return (
    <div style={{ position: 'relative', width: '100%' }}>
      <InputWrapper className="input" isSecondary={isSecondary}>
        <img src={SearchIcon} alt="Input search logo" />
        <StyledInput
          isSecondary={isSecondary}
          isSugestionsOpen={isSugestionsOpen}
          placeholder={placeholder}
          name={name}
          onChange={(event) => onChange && onChange(event.target.value)}
        />
        <label htmlFor={name}>{label}</label>
      </InputWrapper>
      {isSugestionsOpen && sugestions && sugestions?.length > 1 && (
        <SugestionsList
          isSecondary={isSecondary}
          style={{ position: 'absolute', width: '100%' }}
        >
          {sugestions.map((sugestion) => (
            <li key={sugestion.id}>
              {typeof sugestion.id === 'number'   ? (
                <SugestionButtonItem to={`/hero/${sugestion.id}`}>
                  {sugestion.name}
                </SugestionButtonItem>
              ) : (
                <SugestionItemStyleSpan>
                  {sugestion.name}
                </SugestionItemStyleSpan>
              )}
            </li>
          ))}
        </SugestionsList>
      )}
    </div>
  );
}
