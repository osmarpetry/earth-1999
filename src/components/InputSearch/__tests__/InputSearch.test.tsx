import React from 'react'
import '@testing-library/jest-dom'
import { cleanup, fireEvent, render } from '@testing-library/react'

import renderWithRouter from 'core/utils/tests/renderWithRouter'

import InputSearch from 'components/InputSearch'

afterEach(cleanup)

it('should match InputSearch with sugestions open snapshot', () => {
  expect(
    renderWithRouter(
      <InputSearch
        placeholder='Type something...'
        label="Hero's name"
        isSugestionsOpen={true}
        sugestions={[
          { id: undefined, name: 'hul' },
          { id: 1, name: 'Hulk' },
          { id: 1, name: 'Spider-Man' }
        ]}
        value=''
        onChange={jest.fn()}
      />
    )
  ).toMatchSnapshot()
})


it('should match InputSearch with sugestions close snapshot', () => {
  expect(
    renderWithRouter(
      <InputSearch
        placeholder='Type something...'
        label="Hero's name"
        isSugestionsOpen={true}
        sugestions={[
          { id: undefined, name: 'hul' },
          { id: 1, name: 'Hulk' },
          { id: 1, name: 'Spider-Man' }
        ]}
        value=''
        onChange={jest.fn()}
      />
    )
  ).toMatchSnapshot()
})


it('should trigger the onChange event when is typing something', () => {
  const onChange = jest.fn()
  const { getByLabelText } = render(
    <InputSearch
      placeholder='Type something...'
      label="Hero's name"
      value=''
      onChange={onChange}
    />
  )

  expect(onChange).not.toBeCalled()
  fireEvent.change(getByLabelText("Hero's name"), { target: { value: 'Hulk' } })
  expect(onChange).toBeCalledTimes(1)
})

it('should show sugestions list', () => {
  const onChange = jest.fn()
  const { getByText } = renderWithRouter(
    <InputSearch
      placeholder='Type something...'
      label="Hero's name"
      isSugestionsOpen={true}
      sugestions={[
        { id: undefined, name: 'hul' },
        { id: 1, name: 'Hulk' },
        { id: 1, name: 'Spider-Man' }
      ]}
      value=''
      onChange={onChange}
    />
  )
  expect(getByText('hul')).toBeTruthy()
  expect(getByText('Hulk')).toBeTruthy()
  expect(getByText('Hulk')).toBeTruthy()
})
