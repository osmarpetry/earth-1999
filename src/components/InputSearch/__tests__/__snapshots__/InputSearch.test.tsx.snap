// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match InputSearch with sugestions close snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="position: relative; width: 100%;"
      >
        <div
          class="sc-bdfBwQ ikLnIo input"
        >
          <img
            alt="Input search logo"
            src="SearchIcon.svg"
          />
          <input
            class="sc-gsTCUz HbFzO"
            id="search"
            name="search"
            placeholder="Type something..."
          />
          <label
            for="search"
          >
            Hero's name
          </label>
        </div>
        <ul
          class="sc-dlfnbm duLtPO"
          style="position: absolute; width: 100%;"
        >
          <li>
            <span
              class="sc-eCssSg cGrfJp"
            >
              hul
            </span>
          </li>
          <li>
            <a
              class="sc-hKgILt gARJzy"
              data-testid="route"
              href="/hero/1"
            >
              Hulk
            </a>
          </li>
          <li>
            <a
              class="sc-hKgILt gARJzy"
              data-testid="route"
              href="/hero/1"
            >
              Spider-Man
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="position: relative; width: 100%;"
    >
      <div
        class="sc-bdfBwQ ikLnIo input"
      >
        <img
          alt="Input search logo"
          src="SearchIcon.svg"
        />
        <input
          class="sc-gsTCUz HbFzO"
          id="search"
          name="search"
          placeholder="Type something..."
        />
        <label
          for="search"
        >
          Hero's name
        </label>
      </div>
      <ul
        class="sc-dlfnbm duLtPO"
        style="position: absolute; width: 100%;"
      >
        <li>
          <span
            class="sc-eCssSg cGrfJp"
          >
            hul
          </span>
        </li>
        <li>
          <a
            class="sc-hKgILt gARJzy"
            data-testid="route"
            href="/hero/1"
          >
            Hulk
          </a>
        </li>
        <li>
          <a
            class="sc-hKgILt gARJzy"
            data-testid="route"
            href="/hero/1"
          >
            Spider-Man
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should match InputSearch with sugestions open snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="position: relative; width: 100%;"
      >
        <div
          class="sc-bdfBwQ ikLnIo input"
        >
          <img
            alt="Input search logo"
            src="SearchIcon.svg"
          />
          <input
            class="sc-gsTCUz HbFzO"
            id="search"
            name="search"
            placeholder="Type something..."
          />
          <label
            for="search"
          >
            Hero's name
          </label>
        </div>
        <ul
          class="sc-dlfnbm duLtPO"
          style="position: absolute; width: 100%;"
        >
          <li>
            <span
              class="sc-eCssSg cGrfJp"
            >
              hul
            </span>
          </li>
          <li>
            <a
              class="sc-hKgILt gARJzy"
              data-testid="route"
              href="/hero/1"
            >
              Hulk
            </a>
          </li>
          <li>
            <a
              class="sc-hKgILt gARJzy"
              data-testid="route"
              href="/hero/1"
            >
              Spider-Man
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="position: relative; width: 100%;"
    >
      <div
        class="sc-bdfBwQ ikLnIo input"
      >
        <img
          alt="Input search logo"
          src="SearchIcon.svg"
        />
        <input
          class="sc-gsTCUz HbFzO"
          id="search"
          name="search"
          placeholder="Type something..."
        />
        <label
          for="search"
        >
          Hero's name
        </label>
      </div>
      <ul
        class="sc-dlfnbm duLtPO"
        style="position: absolute; width: 100%;"
      >
        <li>
          <span
            class="sc-eCssSg cGrfJp"
          >
            hul
          </span>
        </li>
        <li>
          <a
            class="sc-hKgILt gARJzy"
            data-testid="route"
            href="/hero/1"
          >
            Hulk
          </a>
        </li>
        <li>
          <a
            class="sc-hKgILt gARJzy"
            data-testid="route"
            href="/hero/1"
          >
            Spider-Man
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
